File: dotnet/src/HybridRow/Layouts/LayoutCompiler.cs
// ------------------------------------------------------------
//  Copyright (c) Microsoft Corporation.  All rights reserved.
// ------------------------------------------------------------

namespace Microsoft.Azure.Cosmos.Serialization.HybridRow.Layouts
{
    using System.Collections.Generic;
    using Microsoft.Azure.Cosmos.Core;
    using Microsoft.Azure.Cosmos.Serialization.HybridRow.Schemas;

    /// <summary>Converts a logical schema into a physical layout.</summary>
    internal sealed class LayoutCompiler
    {
        /// <summary>Compiles a logical schema into a physical layout that can be used to read and write rows.</summary>
        /// <param name="ns">The namespace within which <paramref name="schema" /> is defined.</param>
        /// <param name="schema">The logical schema to produce a layout for.</param>
        /// <returns>The layout for the schema.</returns>
        public static Layout Compile(Namespace ns, Schema schema)
        {
            Contract.Requires(ns != null);
            Contract.Requires(schema != null);
            Contract.Requires(schema.Type == TypeKind.Schema);
            Contract.Requires(!string.IsNullOrWhiteSpace(schema.Name));
            Contract.Requires(ns.Schemas.Contains(schema));

            LayoutBuilder builder = new LayoutBuilder(schema.Name, schema.SchemaId);
            LayoutCompiler.AddProperties(builder, ns, LayoutCode.Schema, schema.Properties);

            return builder.Build();
        }

        private static void AddProperties(LayoutBuilder builder, Namespace ns, LayoutCode scope, List<Property> properties)
        {
            foreach (Property p in properties)
            {
                TypeArgumentList typeArgs;
                LayoutType type = LayoutCompiler.LogicalToPhysicalType(ns, p.PropertyType, out typeArgs);
                switch (LayoutCodeTraits.ClearImmutableBit(type.LayoutCode))
                {
                    case LayoutCode.ObjectScope:
                    {
                        if (!p.PropertyType.Nullable)
                        {
                            throw new LayoutCompilationException("Non-nullable sparse column are not supported.");
                        }

                        ObjectPropertyType op = (ObjectPropertyType)p.PropertyType;
                        builder.AddObjectScope(p.Path, type);
                        LayoutCompiler.AddProperties(builder, ns, type.LayoutCode, op.Properties);
                        builder.EndObjectScope();
                        break;
                    }

                    case LayoutCode.ArrayScope:
                    case LayoutCode.TypedArrayScope:
                    case LayoutCode.SetScope:
                    case LayoutCode.TypedSetScope:
                    case LayoutCode.MapScope:
                    case LayoutCode.TypedMapScope:
                    case LayoutCode.TupleScope:
                    case LayoutCode.TypedTupleScope:
                    case LayoutCode.TaggedScope:
                    case LayoutCode.Tagged2Scope:
                    case LayoutCode.Schema:
                    {
                        if (!p.PropertyType.Nullable)
                        {
                            throw new LayoutCompilationException("Non-nullable sparse column are not supported.");
                        }

                        builder.AddTypedScope(p.Path, type, typeArgs);
                        break;
                    }

                    case LayoutCode.NullableScope:
                    {
                        throw new LayoutCompilationException("Nullables cannot be explicitly declared as columns.");
                    }

                    default:
                    {
                        PrimitivePropertyType pp = p.PropertyType as PrimitivePropertyType;
                        if (pp != null)
                        {
                            switch (pp.Storage)
                            {
                                case StorageKind.Fixed:
                                    if (LayoutCodeTraits.ClearImmutableBit(scope) != LayoutCode.Schema)
                                    {
                                        throw new LayoutCompilationException("Cannot have fixed storage within a sparse scope.");
                                    }

                                    if (type.IsNull && !pp.Nullable)
                                    {
                                        throw new LayoutCompilationException("Non-nullable null columns are not supported.");
                                    }

                                    builder.AddFixedColumn(p.Path, type, pp.Nullable, pp.Length);
                                    break;
                                case StorageKind.Variable:
                                    if (LayoutCodeTraits.ClearImmutableBit(scope) != LayoutCode.Schema)
                                    {
                                        throw new LayoutCompilationException("Cannot have variable storage within a sparse scope.");
                                    }

                                    if (!pp.Nullable)
                                    {
                                        throw new LayoutCompilationException("Non-nullable variable columns are not supported.");
                                    }

                                    builder.AddVariableColumn(p.Path, type, pp.Length);
                                    break;
                                case StorageKind.Sparse:
                                    if (!pp.Nullable)
                                    {
                                        throw new LayoutCompilationException("Non-nullable sparse columns are not supported.");
                                    }

                                    builder.AddSparseColumn(p.Path, type);
                                    break;
                                default:
                                    throw new LayoutCompilationException($"Unknown storage specification: {pp.Storage}");
                            }
                        }
                        else
                        {
                            throw new LayoutCompilationException($"Unknown property type: {type.Name}");
                        }

                        break;
                    }
                }
            }
        }

        private static LayoutType LogicalToPhysicalType(Namespace ns, PropertyType logicalType, out TypeArgumentList typeArgs)
        {
            typeArgs = TypeArgumentList.Empty;
            bool immutable = (logicalType as ScopePropertyType)?.Immutable ?? false;

            switch (logicalType.Type)
            {
                case TypeKind.Null:
                    return LayoutType.Null;
                case TypeKind.Boolean:
                    return LayoutType.Boolean;
                case TypeKind.Int8:
                    return LayoutType.Int8;
                case TypeKind.Int16:
                    return LayoutType.Int16;
                case TypeKind.Int32:
                    return LayoutType.Int32;
                case TypeKind.Int64:
                    return LayoutType.Int64;
                case TypeKind.UInt8:
                    return LayoutType.UInt8;
                case TypeKind.UInt16:
                    return LayoutType.UInt16;
                case TypeKind.UInt32:
                    return LayoutType.UInt32;
                case TypeKind.UInt64:
                    return LayoutType.UInt64;
                case TypeKind.Float32:
                    return LayoutType.Float32;
                case TypeKind.Float64:
                    return LayoutType.Float64;
                case TypeKind.Float128:
                    return LayoutType.Float128;
                case TypeKind.Decimal:
                    return LayoutType.Decimal;
                case TypeKind.DateTime:
                    return LayoutType.DateTime;
                case TypeKind.UnixDateTime:
                    return LayoutType.UnixDateTime;
                case TypeKind.Guid:
                    return LayoutType.Guid;
                case TypeKind.MongoDbObjectId:
                    return LayoutType.MongoDbObjectId;
                case TypeKind.Utf8:
                    return LayoutType.Utf8;
                case TypeKind.Binary:
                    return LayoutType.Binary;
                case TypeKind.VarInt:
                    return LayoutType.VarInt;
                case TypeKind.VarUInt:
                    return LayoutType.VarUInt;

                case TypeKind.Object:
                    return immutable ? LayoutType.ImmutableObject : LayoutType.Object;
                case TypeKind.Array:
                    ArrayPropertyType ap = (ArrayPropertyType)logicalType;
                    if ((ap.Items != null) && (ap.Items.Type != TypeKind.Any))
                    {
                        TypeArgumentList itemTypeArgs;
                        LayoutType itemType = LayoutCompiler.LogicalToPhysicalType(ns, ap.Items, out itemTypeArgs);
                        if (ap.Items.Nullable)
                        {
                            itemTypeArgs = new TypeArgumentList(new[] { new TypeArgument(itemType, itemTypeArgs) });
                            itemType = itemType.Immutable ? LayoutType.ImmutableNullable : LayoutType.Nullable;
                        }

                        typeArgs = new TypeArgumentList(new[] { new TypeArgument(itemType, itemTypeArgs) });
                        return immutable ? LayoutType.ImmutableTypedArray : LayoutType.TypedArray;
                    }

                    return immutable ? LayoutType.ImmutableArray : LayoutType.Array;
                case TypeKind.Set:
                    SetPropertyType sp = (SetPropertyType)logicalType;
                    if ((sp.Items != null) && (sp.Items.Type != TypeKind.Any))
                    {
                        TypeArgumentList itemTypeArgs;
                        LayoutType itemType = LayoutCompiler.LogicalToPhysicalType(ns, sp.Items, out itemTypeArgs);
                        if (sp.Items.Nullable)
                        {
                            itemTypeArgs = new TypeArgumentList(new[] { new TypeArgument(itemType, itemTypeArgs) });
                            itemType = itemType.Immutable ? LayoutType.ImmutableNullable : LayoutType.Nullable;
                        }

                        typeArgs = new TypeArgumentList(new[] { new TypeArgument(itemType, itemTypeArgs) });
                        return immutable ? LayoutType.ImmutableTypedSet : LayoutType.TypedSet;
                    }

                    // TODO(283638): implement sparse set.
                    throw new LayoutCompilationException($"Unknown property type: {logicalType.Type}");

                case TypeKind.Map:
                    MapPropertyType mp = (MapPropertyType)logicalType;
                    if ((mp.Keys != null) && (mp.Keys.Type != TypeKind.Any) && (mp.Values != null) && (mp.Values.Type != TypeKind.Any))
                    {
                        TypeArgumentList keyTypeArgs;
                        LayoutType keyType = LayoutCompiler.LogicalToPhysicalType(ns, mp.Keys, out keyTypeArgs);
                        if (mp.Keys.Nullable)
                        {
                            keyTypeArgs = new TypeArgumentList(new[] { new TypeArgument(keyType, keyTypeArgs) });
                            keyType = keyType.Immutable ? LayoutType.ImmutableNullable : LayoutType.Nullable;
                        }

                        TypeArgumentList valueTypeArgs;
                        LayoutType valueType = LayoutCompiler.LogicalToPhysicalType(ns, mp.Values, out valueTypeArgs);
                        if (mp.Values.Nullable)
                        {
                            valueTypeArgs = new TypeArgumentList(new[] { new TypeArgument(valueType, valueTypeArgs) });
                            valueType = valueType.Immutable ? LayoutType.ImmutableNullable : LayoutType.Nullable;
                        }

                        typeArgs = new TypeArgumentList(new[] { new TypeArgument(keyType, keyTypeArgs), new TypeArgument(valueType, valueTypeArgs) });
                        return immutable ? LayoutType.ImmutableTypedMap : LayoutType.TypedMap;
                    }

                    // TODO(283638): implement sparse map.
                    throw new LayoutCompilationException($"Unknown property type: {logicalType.Type}");

                case TypeKind.Tuple:
                    TuplePropertyType tp = (TuplePropertyType)logicalType;
                    TypeArgument[] args = new TypeArgument[tp.Items.Count];
                    for (int i = 0; i < tp.Items.Count; i++)
                    {
                        TypeArgumentList itemTypeArgs;
                        LayoutType itemType = LayoutCompiler.LogicalToPhysicalType(ns, tp.Items[i], out itemTypeArgs);
                        if (tp.Items[i].Nullable)
                        {
                            itemTypeArgs = new TypeArgumentList(new[] { new TypeArgument(itemType, itemTypeArgs) });
                            itemType = itemType.Immutable ? LayoutType.ImmutableNullable : LayoutType.Nullable;
                        }

                        args[i] = new TypeArgument(itemType, itemTypeArgs);
                    }

                    typeArgs = new TypeArgumentList(args);
                    return immutable ? LayoutType.ImmutableTypedTuple : LayoutType.TypedTuple;

                case TypeKind.Tagged:
                    TaggedPropertyType tg = (TaggedPropertyType)logicalType;
                    if ((tg.Items.Count < TaggedPropertyType.MinTaggedArguments) || (tg.Items.Count > TaggedPropertyType.MaxTaggedArguments))
                    {
                        throw new LayoutCompilationException(
                            $"Invalid number of arguments in Tagged: {TaggedPropertyType.MinTaggedArguments} <= {tg.Items.Count} <= {TaggedPropertyType.MaxTaggedArguments}");
                    }

                    TypeArgument[] tgArgs = new TypeArgument[tg.Items.Count + 1];
                    tgArgs[0] = new TypeArgument(LayoutType.UInt8, TypeArgumentList.Empty);
                    for (int i = 0; i < tg.Items.Count; i++)
                    {
                        TypeArgumentList itemTypeArgs;
                        LayoutType itemType = LayoutCompiler.LogicalToPhysicalType(ns, tg.Items[i], out itemTypeArgs);
                        if (tg.Items[i].Nullable)
                        {
                            itemTypeArgs = new TypeArgumentList(new[] { new TypeArgument(itemType, itemTypeArgs) });
                            itemType = itemType.Immutable ? LayoutType.ImmutableNullable : LayoutType.Nullable;
                        }

                        tgArgs[i + 1] = new TypeArgument(itemType, itemTypeArgs);
                    }

                    typeArgs = new TypeArgumentList(tgArgs);
                    switch (tg.Items.Count)
                    {
                        case 1:
                            return immutable ? LayoutType.ImmutableTagged : LayoutType.Tagged;
                        case 2:
                            return immutable ? LayoutType.ImmutableTagged2 : LayoutType.Tagged2;
                        default:
                            throw new LayoutCompilationException("Unexpected tagged arity");
                    }

                case TypeKind.Schema:
                    UdtPropertyType up = (UdtPropertyType)logicalType;
                    Schema udtSchema;
                    if (up.SchemaId == SchemaId.Invalid)
                    {
                        udtSchema = ns.Schemas.Find(s => s.Name == up.Name);
                    }
                    else
                    {
                        udtSchema = ns.Schemas.Find(s => s.SchemaId == up.SchemaId);
                        if (udtSchema.Name != up.Name)
                        {
                            throw new LayoutCompilationException($"Ambiguous schema reference: '{up.Name}:{up.SchemaId}'");
                        }
                    }

                    if (udtSchema == null)
                    {
                        throw new LayoutCompilationException($"Cannot resolve schema reference '{up.Name}:{up.SchemaId}'");
                    }

                    typeArgs = new TypeArgumentList(udtSchema.SchemaId);
                    return immutable ? LayoutType.ImmutableUDT : LayoutType.UDT;

                default:
                    throw new LayoutCompilationException($"Unknown property type: {logicalType.Type}");
            }
        }
    }
}

File: java/src/main/java/com/azure/data/cosmos/serialization/hybridrow/layouts/LayoutCompiler.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.azure.data.cosmos.serialization.hybridrow.layouts;

import com.azure.data.cosmos.core.Out;
import com.azure.data.cosmos.serialization.hybridrow.SchemaId;
import com.azure.data.cosmos.serialization.hybridrow.schemas.ArrayPropertyType;
import com.azure.data.cosmos.serialization.hybridrow.schemas.MapPropertyType;
import com.azure.data.cosmos.serialization.hybridrow.schemas.Namespace;
import com.azure.data.cosmos.serialization.hybridrow.schemas.ObjectPropertyType;
import com.azure.data.cosmos.serialization.hybridrow.schemas.PrimitivePropertyType;
import com.azure.data.cosmos.serialization.hybridrow.schemas.Property;
import com.azure.data.cosmos.serialization.hybridrow.schemas.PropertyType;
import com.azure.data.cosmos.serialization.hybridrow.schemas.Schema;
import com.azure.data.cosmos.serialization.hybridrow.schemas.ScopePropertyType;
import com.azure.data.cosmos.serialization.hybridrow.schemas.SetPropertyType;
import com.azure.data.cosmos.serialization.hybridrow.schemas.TaggedPropertyType;
import com.azure.data.cosmos.serialization.hybridrow.schemas.TuplePropertyType;
import com.azure.data.cosmos.serialization.hybridrow.schemas.TypeKind;
import com.azure.data.cosmos.serialization.hybridrow.schemas.UdtPropertyType;
import tangible.ListHelper;

import java.util.List;

import static com.google.common.base.Preconditions.checkArgument;

/**
 * Converts a logical schema into a physical layout.
 */
public final class LayoutCompiler {
    /**
     * Compiles a logical schema into a physical layout that can be used to read and write rows.
     *
     * @param ns     The namespace within which <paramref name="schema" /> is defined.
     * @param schema The logical schema to produce a layout for.
     * @return The layout for the schema.
     */
    public static Layout Compile(Namespace ns, Schema schema) {
        checkArgument(ns != null);
        checkArgument(schema != null);
        checkArgument(schema.type() == TypeKind.Schema);
        checkArgument(!tangible.StringHelper.isNullOrWhiteSpace(schema.name()));
        checkArgument(ns.schemas().contains(schema));

        LayoutBuilder builder = new LayoutBuilder(schema.name(), schema.schemaId());
        LayoutCompiler.addProperties(builder, ns, LayoutCode.SCHEMA, schema.properties());

        return builder.build();
    }

    private static void addProperties(LayoutBuilder builder, Namespace ns, LayoutCode scope, List<Property> properties) {

        final Out<TypeArgumentList> typeArgs = new Out<>();

        for (Property p : properties) {

            LayoutType type = LayoutCompiler.logicalToPhysicalType(ns, p.propertyType(), typeArgs);

            switch (LayoutCodeTraits.clearImmutableBit(type.layoutCode())) {

                case OBJECT_SCOPE: {
                    if (!p.propertyType().nullable()) {
                        throw new LayoutCompilationException("Non-nullable sparse column are not supported.");
                    }

                    ObjectPropertyType op = (ObjectPropertyType)p.propertyType();
                    builder.addObjectScope(p.path(), type);
                    LayoutCompiler.addProperties(builder, ns, type.layoutCode(), op.properties());
                    builder.EndObjectScope();
                    break;
                }

                case ARRAY_SCOPE:
                case TYPED_ARRAY_SCOPE:
                case SET_SCOPE:
                case TYPED_SET_SCOPE:
                case MAP_SCOPE:
                case TYPED_MAP_SCOPE:
                case TUPLE_SCOPE:
                case TYPED_TUPLE_SCOPE:
                case TAGGED_SCOPE:
                case TAGGED2_SCOPE:
                case SCHEMA: {
                    if (!p.propertyType().nullable()) {
                        throw new LayoutCompilationException("Non-nullable sparse column are not supported.");
                    }
                    builder.addTypedScope(p.path(), type, typeArgs.get());
                    break;
                }

                case NULLABLE_SCOPE: {
                    throw new LayoutCompilationException("Nullables cannot be explicitly declared as columns.");
                }

                default: {
                    PropertyType tempVar = p.propertyType();
                    PrimitivePropertyType pp = tempVar instanceof PrimitivePropertyType ?
                        (PrimitivePropertyType)tempVar : null;
                    if (pp != null) {
                        switch (pp.storage()) {
                            case FIXED:
                                if (LayoutCodeTraits.clearImmutableBit(scope) != LayoutCode.SCHEMA) {
                                    throw new LayoutCompilationException("Cannot have fixed storage within a sparse " +
                                        "scope.");
                                }

                                if (type.isNull() && !pp.nullable()) {
                                    throw new LayoutCompilationException("Non-nullable null columns are not supported" +
                                        ".");
                                }

                                builder.addFixedColumn(p.path(), type, pp.nullable(), pp.length());
                                break;
                            case VARIABLE:
                                if (LayoutCodeTraits.clearImmutableBit(scope) != LayoutCode.SCHEMA) {
                                    throw new LayoutCompilationException("Cannot have variable storage within a " +
                                        "sparse scope.");
                                }

                                if (!pp.nullable()) {
                                    throw new LayoutCompilationException("Non-nullable variable columns are not " +
                                        "supported.");
                                }

                                builder.addVariableColumn(p.path(), type, pp.length());
                                break;
                            case SPARSE:
                                if (!pp.nullable()) {
                                    throw new LayoutCompilationException("Non-nullable sparse columns are not " +
                                        "supported.");
                                }

                                builder.addSparseColumn(p.path(), type);
                                break;
                            default:
                                throw new LayoutCompilationException(String.format("Unknown storage specification: " +
                                    "%1$s", pp.storage()));
                        }
                    } else {
                        throw new LayoutCompilationException(String.format("Unknown property type: %1$s",
                            type.name()));
                    }

                    break;
                }
            }
        }
    }

    private static LayoutType logicalToPhysicalType(Namespace ns, PropertyType logicalType,
                                                    Out<TypeArgumentList> typeArgs) {
        typeArgs.setAndGet(TypeArgumentList.EMPTY);
        boolean tempVar =
            (logicalType instanceof ScopePropertyType ? (ScopePropertyType)logicalType : null).immutable();
        boolean immutable =
            (logicalType instanceof ScopePropertyType ? (ScopePropertyType)logicalType : null) == null ? null :
                tempVar != null && tempVar;

        switch (logicalType.type()) {
            case Null:
                return LayoutTypes.NULL;
            case Boolean:
                return LayoutTypes.BOOLEAN;
            case Int8:
                return LayoutTypes.INT_8;
            case Int16:
                return LayoutTypes.INT_16;
            case Int32:
                return LayoutTypes.INT_32;
            case Int64:
                return LayoutTypes.INT_64;
            case UInt8:
                return LayoutTypes.UINT_8;
            case UInt16:
                return LayoutTypes.UINT_16;
            case UInt32:
                return LayoutTypes.UINT_32;
            case UInt64:
                return LayoutTypes.UINT_64;
            case Float32:
                return LayoutTypes.FLOAT_32;
            case Float64:
                return LayoutTypes.FLOAT_64;
            case Float128:
                return LayoutTypes.FLOAT_128;
            case Decimal:
                return LayoutTypes.DECIMAL;
            case DateTime:
                return LayoutTypes.DATE_TIME;
            case UnixDateTime:
                return LayoutTypes.UNIX_DATE_TIME;
            case Guid:
                return LayoutTypes.GUID;
            case MongoDbObjectId:
                throw new UnsupportedOperationException();
                // return LayoutTypes.MONGO_DB_OBJECT_ID;
            case Utf8:
                return LayoutTypes.UTF_8;
            case Binary:
                return LayoutTypes.BINARY;
            case VarInt:
                return LayoutTypes.VAR_INT;
            case VarUInt:
                return LayoutTypes.VAR_UINT;

            case Object:
                return immutable ? LayoutTypes.IMMUTABLE_OBJECT : LayoutTypes.OBJECT;
            case Array:
                ArrayPropertyType ap = (ArrayPropertyType)logicalType;
                if ((ap.items() != null) && (ap.items().type() != TypeKind.Any)) {
                    TypeArgumentList itemTypeArgs = new TypeArgumentList();
                    Out<TypeArgumentList> tempOut_itemTypeArgs = new Out<TypeArgumentList>();
                    LayoutType itemType = LayoutCompiler.logicalToPhysicalType(ns, ap.items(), tempOut_itemTypeArgs);
                    itemTypeArgs = tempOut_itemTypeArgs.get();
                    if (ap.items().nullable()) {
                        itemTypeArgs = new TypeArgumentList(new TypeArgument[] { new TypeArgument(itemType,
                            itemTypeArgs) });
                        itemType = itemType.isImmutable() ? LayoutTypes.IMMUTABLE_NULLABLE : LayoutTypes.NULLABLE;
                    }

                    typeArgs.setAndGet(new TypeArgumentList(new TypeArgument[] { new TypeArgument(itemType,
                        itemTypeArgs) }));
                    return immutable ? LayoutTypes.IMMUTABLE_TYPED_ARRAY : LayoutTypes.TYPED_ARRAY;
                }

                return immutable ? LayoutTypes.IMMUTABLE_ARRAY : LayoutTypes.ARRAY;
            case SET:
                SetPropertyType sp = (SetPropertyType)logicalType;
                if ((sp.items() != null) && (sp.items().type() != TypeKind.Any)) {
                    TypeArgumentList itemTypeArgs = new TypeArgumentList();
                    Out<TypeArgumentList> tempOut_itemTypeArgs2 = new Out<TypeArgumentList>();
                    LayoutType itemType = LayoutCompiler.logicalToPhysicalType(ns, sp.items(),
                        tempOut_itemTypeArgs2);
                    itemTypeArgs = tempOut_itemTypeArgs2.get();
                    if (sp.items().nullable()) {
                        itemTypeArgs = new TypeArgumentList(new TypeArgument[] { new TypeArgument(itemType,
                            itemTypeArgs) });
                        itemType = itemType.isImmutable() ? LayoutTypes.IMMUTABLE_NULLABLE : LayoutTypes.NULLABLE;
                    }

                    typeArgs.setAndGet(new TypeArgumentList(new TypeArgument[] { new TypeArgument(itemType,
                        itemTypeArgs) }));
                    return immutable ? LayoutTypes.IMMUTABLE_TYPED_SET : LayoutTypes.TYPED_SET;
                }

                // TODO(283638): implement sparse set.
                throw new LayoutCompilationException(String.format("Unknown property type: %1$s",
                    logicalType.type()));

            case MAP:
                MapPropertyType mp = (MapPropertyType)logicalType;
                if ((mp.keys() != null) && (mp.keys().type() != TypeKind.Any) && (mp.values() != null) && (mp.values().type() != TypeKind.Any)) {
                    TypeArgumentList keyTypeArgs = new TypeArgumentList();
                    Out<TypeArgumentList> tempOut_keyTypeArgs = new Out<TypeArgumentList>();
                    LayoutType keyType = LayoutCompiler.logicalToPhysicalType(ns, mp.keys(), tempOut_keyTypeArgs);
                    keyTypeArgs = tempOut_keyTypeArgs.get();
                    if (mp.keys().nullable()) {
                        keyTypeArgs = new TypeArgumentList(new TypeArgument[] { new TypeArgument(keyType,
                            keyTypeArgs) });
                        keyType = keyType.isImmutable() ? LayoutTypes.IMMUTABLE_NULLABLE : LayoutTypes.NULLABLE;
                    }

                    TypeArgumentList valueTypeArgs = new TypeArgumentList();
                    Out<TypeArgumentList> tempOut_valueTypeArgs = new Out<TypeArgumentList>();
                    LayoutType valueType = LayoutCompiler.logicalToPhysicalType(ns, mp.values(),
                        tempOut_valueTypeArgs);
                    valueTypeArgs = tempOut_valueTypeArgs.get();
                    if (mp.values().nullable()) {
                        valueTypeArgs = new TypeArgumentList(new TypeArgument[] { new TypeArgument(valueType,
                            valueTypeArgs) });
                        valueType = valueType.isImmutable() ? LayoutTypes.IMMUTABLE_NULLABLE : LayoutTypes.NULLABLE;
                    }

                    typeArgs.setAndGet(new TypeArgumentList(new TypeArgument[]
                        {
                            new TypeArgument(keyType, keyTypeArgs),
                            new TypeArgument(valueType, valueTypeArgs)
                        }));
                    return immutable ? LayoutTypes.IMMUTABLE_TYPED_MAP : LayoutTypes.TYPED_MAP;
                }

                // TODO(283638): implement sparse map.
                throw new LayoutCompilationException(String.format("Unknown property type: %1$s",
                    logicalType.type()));

            case Tuple:
                TuplePropertyType tp = (TuplePropertyType)logicalType;
                TypeArgument[] args = new TypeArgument[tp.items().size()];
                for (int i = 0; i < tp.items().size(); i++) {
                    TypeArgumentList itemTypeArgs = new TypeArgumentList();
                    Out<TypeArgumentList> tempOut_itemTypeArgs3 = new Out<TypeArgumentList>();
                    LayoutType itemType = LayoutCompiler.logicalToPhysicalType(ns, tp.items().get(i),
                        tempOut_itemTypeArgs3);
                    itemTypeArgs = tempOut_itemTypeArgs3.get();
                    if (tp.items().get(i).nullable()) {
                        itemTypeArgs = new TypeArgumentList(new TypeArgument[] { new TypeArgument(itemType,
                            itemTypeArgs) });
                        itemType = itemType.isImmutable() ? LayoutTypes.IMMUTABLE_NULLABLE : LayoutTypes.NULLABLE;
                    }

                    args[i] = new TypeArgument(itemType, itemTypeArgs);
                }

                typeArgs.setAndGet(new TypeArgumentList(args));
                return immutable ? LayoutTypes.IMMUTABLE_TYPED_TUPLE : LayoutTypes.TYPED_TUPLE;

            case TAGGED:
                TaggedPropertyType tg = (TaggedPropertyType)logicalType;
                if ((tg.items().size() < TaggedPropertyType.MinTaggedArguments) || (tg.items().size() > TaggedPropertyType.MaxTaggedArguments)) {
                    throw new LayoutCompilationException(String.format("Invalid number of arguments in Tagged: %1$s " +
                        "<= %2$s <= %3$s", TaggedPropertyType.MinTaggedArguments, tg.items().size(),
                        TaggedPropertyType.MaxTaggedArguments));
                }

                TypeArgument[] tgArgs = new TypeArgument[tg.items().size() + 1];
                tgArgs[0] = new TypeArgument(LayoutTypes.UINT_8, TypeArgumentList.EMPTY);
                for (int i = 0; i < tg.items().size(); i++) {
                    TypeArgumentList itemTypeArgs = new TypeArgumentList();
                    Out<TypeArgumentList> tempOut_itemTypeArgs4 = new Out<TypeArgumentList>();
                    LayoutType itemType = LayoutCompiler.logicalToPhysicalType(ns, tg.items().get(i),
                        tempOut_itemTypeArgs4);
                    itemTypeArgs = tempOut_itemTypeArgs4.get();
                    if (tg.items().get(i).nullable()) {
                        itemTypeArgs = new TypeArgumentList(new TypeArgument[] { new TypeArgument(itemType,
                            itemTypeArgs) });
                        itemType = itemType.isImmutable() ? LayoutTypes.IMMUTABLE_NULLABLE : LayoutTypes.NULLABLE;
                    }

                    tgArgs[i + 1] = new TypeArgument(itemType, itemTypeArgs);
                }

                typeArgs.setAndGet(new TypeArgumentList(tgArgs));
                switch (tg.items().size()) {
                    case 1:
                        return immutable ? LayoutTypes.IMMUTABLE_TAGGED : LayoutTypes.TAGGED;
                    case 2:
                        return immutable ? LayoutTypes.IMMUTABLE_TAGGED_2 : LayoutTypes.TAGGED_2;
                    default:
                        throw new LayoutCompilationException("Unexpected tagged arity");
                }

            case Schema:
                UdtPropertyType up = (UdtPropertyType)logicalType;
                Schema udtSchema;
                if (SchemaId.opEquals(up.schemaId(), SchemaId.INVALID)) {
                    udtSchema = ListHelper.find(ns.schemas(), s = up.name().equals( > s.Name))
                } else {
                    udtSchema = ListHelper.find(ns.schemas(), s =
                        SchemaId.opEquals( > s.SchemaId, up.schemaId()))
                    if (!udtSchema.name().equals(up.name())) {
                        throw new LayoutCompilationException(String.format("Ambiguous schema reference: '%1$s:%2$s'",
                            up.name(), up.schemaId()));
                    }
                }

                if (udtSchema == null) {
                    throw new LayoutCompilationException(String.format("Cannot resolve schema reference '%1$s:%2$s'",
                        up.name(), up.schemaId()));
                }

                typeArgs.setAndGet(new TypeArgumentList(udtSchema.schemaId()));
                return immutable ? LayoutTypes.IMMUTABLE_UDT : LayoutTypes.UDT;

            default:
                throw new LayoutCompilationException(String.format("Unknown property type: %1$s",
                    logicalType.type()));
        }
    }
}